subsys/net/lib/lwm2m/lwm2m_rd_client.c

```diff
@@ -1380,8 +1386,16 @@ int lwm2m_rd_client_stop(struct lwm2m_ctx *client_ctx,
 
 	k_mutex_unlock(&client.mutex);
 
-	while (get_sm_state() != ENGINE_IDLE) {
-		k_sleep(K_MSEC(STATE_MACHINE_UPDATE_INTERVAL_MS / 2));
+	while (true) {
+		switch(get_sm_state()) {
+		default:
```

Comment:
Any particular reason to put the `default:` first? I know it's still a valid syntax, but personally, I find it a bit confusing.

Reply:
Besides the style of this construct of the while loop that I actually don't like, I am asking whether we can change this function to non-blocking. I have opened already a discussion

https://github.com/zephyrproject-rtos/zephyr/discussions/54478

The way lwm2m_rd_client_stop is currently implemented caused already some trouble for us in our projects as it requires some pre assumption about behaviour of the implementation of the rd client as well as the link state to be **always** non blocking.

I know that this is kind of off topic here, but I would like to raise my concerns here.

Reply:
I have gone through various PRs regarding this and I think the loop was trying to fix a problem that does not exist anymore.
Previously this function moved state directly to IDLE and there was a possibility of socket handle to leak. In PR https://github.com/zephyrproject-rtos/zephyr/pull/44365 it was changed that state is moved to `DEREGISTERED` which would do the internal cleanup.
I don't know why I added the blocking loop because it looks like `lwm2m_rd_client_start()` does not re-start until we are in `IDLE` state.
So I think it is safe to remove this blocking loop.